### DISCUSSION TITLE
System.Object Problem

### DISCUSSION THREAD

**Discussion by chensem** [vfsfitvnm/frida-il2cpp-bridge#230] [Q&A]
```
    const assem = Il2Cpp.Domain.assembly("Assembly-CSharp").image;
    const object = assem.class("ProtobufPacket.CG_CARD_MAKE").alloc();
    object.method('.ctor').invoke()

    const type = object.tryMethod<Il2Cpp.Object>("GetType").invoke();
    // console.log(type.class) // why System.RuntimeType
    // console.log(type.tryMethod("GetProperties" , 0))
    const properties = type.tryMethod<Il2Cpp.Array>("GetProperties" , 0).invoke();
    const fields = type.tryMethod<Il2Cpp.Array>("GetFields" , 0).invoke();

    for(let property of properties){
        const propertyInfo = property as Il2Cpp.Object // System.Reflection.PropertyInfo 
        console.log(propertyInfo.tryMethod("get_Name").invoke())

        // System.Object GetValue(System.Object obj, System.Object[] index); // 0x0183bf4c
        propertyInfo.tryMethod("GetValue").invoke(object , 0); // args0 ??  
    }

    console.log(properties , fields)
```

> propertyInfo.tryMethod("GetValue").invoke(object , 0); // args0 ??  

get errors like below 

> il2cpp: cannot invoke method GetValue: parameter types mismatch

how to convert from object ?  that the args[0] can be handled well

**Comment by bluewave41** [comment]
`parameter types mismatch`

Your method takes: `(System.Object obj, System.Object[] index)`
You've given:           `object, 0` `(System.Object, int)`

You showed the class in https://github.com/vfsfitvnm/frida-il2cpp-bridge/discussions/231 and there is no `GetValue(System.Object, int)`. Your parameter types don't match.

It's not clear from this what the actual type of the System.Object[] is. It will be something that inherits from System.Object so you should [trace](https://github.com/vfsfitvnm/frida-il2cpp-bridge/wiki/Snippets#tracing) it to determine what it really is.

Then you'll need to construct an object that satisfies it. `Il2Cpp.Array.from(type, value)` will construct an array that you can pass in.

I figure you'll end up with something resembling 
```
const o = assem.class('x').alloc();
o.method('.ctor').invoke();
propertyInfo.tryMethod("GetValue").invoke(object , Il2Cpp.Array.from(assem.class('some type'), [o]);
```

But it all depends on how this is called internally.
